src/gap-skipper.js

```diff
@@ -0,0 +1,163 @@
+/**
+ * the gap skipper object handles all scenarios
+ * where the player runs into the end of a buffered
+ * region and there is a buffered region ahead.
+ * It then handles the skipping behavior.
+ *
+ * @class GapSkipper
+ */
+export default class GapSkipper {
+
+  /**
+  * Represents a GapSKipper object.
+  * @constructor
+  * @param {object} options an object that includes the tech and settings
+  */
+  constructor(options) {
+    if (!options.tech.options_.playerId) {
+      return;
+    }
+
+    this.player = videojs(options.tech.options_.playerId);
+    this.tech_ = options.tech;
+    this.consecutiveUpdates = 0;
+    this.timer = null;
+    this.lastRecordedTime = null;
+
+    if (options.debug) {
+      this.logger_ = videojs.log.bind(videojs, '<gap-skipper>');
+    }
+
+    this.player.one('canplaythrough', () => {
+      this.player.on('waiting', () => {
+        this.setTimer_();
+      });
+
+      // The purpose of this function is to emulate the "waiting" event on
+      // browsers that do not emit it when they are stalled waiting for
+      // more data
+      this.player.on('timeupdate', () => {
+        if (this.player.paused()) {
+          return;
+        }
+
+        let currentTime = this.player.currentTime();
+
+        if (this.consecutiveUpdates === 5 &&
+            currentTime === this.lastRecordedTime) {
+
+          // trigger waiting
+          this.player.trigger('waiting');
+          this.consecutiveUpdates++;
+        } else if (currentTime === this.lastRecordedTime) {
+          this.consecutiveUpdates++;
+        } else {
+          this.consecutiveUpdates = 0;
+          this.lastRecordedTime = currentTime;
+        }
+      });
+
+      // Set of conditions that reset the gap-skipper logic
+      [
+        'seeking',
+        'seeked',
+        'pause',
+        'playing',
+        'error'
+      ].forEach((event) => {
+        this.player.on(event, () => {
+          this.cancelTimer_();
+        });
+      });
+    });
+  }
+
+  /**
+  * Cancels any pending timers and resets the 'timeupdate' mechanism
+  * designed to detect that we are stalled
+  *
+  * @private
+  */
+  cancelTimer_() {
+    this.consecutiveUpdates = 0;
+
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+
+    this.timer = null;
+  }
+
+  /**
+  * Timer callback. If playback still has not proceeded, then we seek
+  * to the start of the next buffered region.
+  *
+  * @private
+  */
+  skipTheGap_(scheduledCurrentTime) {
+    let buffered = this.player.buffered();
+    let currentTime = this.player.currentTime();
+    let nextRange = Ranges.findNextRange(buffered, currentTime);
+
+    this.consecutiveUpdates = 0;
+    this.timer = null;
+
+    this.logger_('timer triggered');
+
+    if (nextRange.length === 0) {
+      return;
+    }
+
+    this.logger_('currentTime:', currentTime, 'scheduled currentTime:', scheduledCurrentTime, 'nextRange start:', nextRange.start(0));
+
+    if (currentTime !== scheduledCurrentTime) {
```

Comment:
maybe merge this if statement and the one above it since they do the same thing?